### PR TITLE
Fix/get lowest average cost for fruits and vegetables

### PR DESCRIPTION
**Pull request by HaidyAbuGom3a**

### CODE REVIEW

File: src/main/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractor.kt [hunk 1]
```diff
@@ -9,25 +9,27 @@ class GetLowestAveragePricesForFruitsAndVegetablesInteractor (
...
-                city.fruitAndVegetablesPrices.potato1kg!! +
-                city.fruitAndVegetablesPrices.onion1kg!! +
-                city.fruitAndVegetablesPrices.lettuceOneHead!!
-                )/city.averageMonthlyNetSalaryAfterTax!!
-    }
 
-    private fun excludeNullSalariesAndLowQualityDataAndNullFruitsAndVegetablesPrices(city: CityEntity): Boolean {
-        return city.averageMonthlyNetSalaryAfterTax != null &&
+    private fun excludeNullSalariesAndLowQualityDataAndNullFruitsAndVegetablesPrices(city: CityEntity) =
+                city.averageMonthlyNetSalaryAfterTax != null &&
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractor.kt:32]: Use **isNotNull()** extension function for checking null values

File: src/test/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractorTest.kt [hunk 2]
```diff
@@ -22,58 +22,64 @@ internal class GetLowestAveragePricesForFruitsAndVegetablesInteractorTest {
     fun `should return empty list when the limit is equal to zero`() {
```
Comment (ma28alhams) [inline src/test/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractorTest.kt:22]: It should return **InvalidLimitException**

File: src/main/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractor.kt [hunk 3]
```diff
@@ -9,25 +9,27 @@ class GetLowestAveragePricesForFruitsAndVegetablesInteractor (
         return  data
             .getAllCitiesData()
             .filter(::excludeNullSalariesAndLowQualityDataAndNullFruitsAndVegetablesPrices)
-            .sortedBy {getRatio(it)}
-            .map { it.cityName }
+            .sortedBy(::getPricesToSalaryAverage)
             .take(limit)
+            .map { it.cityName }
+
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractor.kt:15]:
```suggestion
    fun execute(limit: Int): List<String> {
        if (limit < 0) throw InvalidLimitException("Limit cannot be zero or negative")
        return  data
            .getAllCitiesData()
            .ifEmpty { throw IllegalStateException("Something went wrong") }
            .filter(::excludeNullSalariesAndLowQualityDataAndNullFruitsAndVegetablesPrices)
            .sortedBy(::getPricesToSalaryAverage)
            .take(limit)
            .map { it.cityName }
    }
```
Then change the test cases.

**Review by ma28alhams — CHANGES_REQUESTED** [review]

File: src/main/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractor.kt [hunk 4]
```diff
@@ -1,40 +1,47 @@
...
         return  data
             .getAllCitiesData()
             .filter(::excludeNullSalariesAndLowQualityDataAndNullFruitsAndVegetablesPrices)
-            .sortedBy {getRatio(it)}
-            .map { it.cityName }
+            .ifEmpty { throw IllegalStateException("Something went wrong, no valid data") }
+            .sortedBy(::getPricesToSalaryAverage)
             .take(limit)
+            .map { it.cityName }
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetLowestAveragePricesForFruitsAndVegetablesInteractor.kt:19]: Remove **ifEmpty**, if the filtered data is empty(that means all data is filtered), then it returns an empty list.
Add a test case if the filtered list is empty then should return empty list.
```suggestion
            .sortedBy(::getPricesToSalaryAverage)
            .take(limit)
            .map { it.cityName }
```

**Review by ma28alhams — CHANGES_REQUESTED** [review]
- Add test case for negative limit(should throw InvalidLimitException)
- Add a test case when the retrieved data from the data source is empty
- Add a test case if the filtered list is empty then should return an empty list.

**Review by ma28alhams — APPROVED** [review]
